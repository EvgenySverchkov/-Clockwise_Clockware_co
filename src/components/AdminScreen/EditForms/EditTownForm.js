import React from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";

import {updateTownInState} from "../../../store/adminPanel/actions";
import putDataToServer from "../services/putDataToServer";

import EditForm from "../../../forms/EditForm";

import { SERVERDOMAIN } from "../../../services/serverUrls";

const EditTownForm = ({match,history}) => {
  const state = useSelector(state=>{
    return {townsArr: state.town_reduser.towns}
  });
  const dispatch = useDispatch();

  function editTownHandler(e, newTownObj) {
    e.preventDefault();
    putDataToServer(`${SERVERDOMAIN}/towns/put/${newTownObj.id}`, newTownObj)
      .then((data) => {
        if (data.success) {
          alert(data.msg);
          dispatch(updateTownInState(data));
          history.push("/admin/townsList");
        } else {
          alert(data.msg);
        }
      })
      .catch((err) => alert(err));
  }
  return (
    <EditForm
      id={+match.params.id}
      handler={editTownHandler}
      arrFromState={state.townsArr}
    />
  );
};

EditTownForm.propTypes = {
  match: PropTypes.object,
  history: PropTypes.object
};

export default EditTownForm;
