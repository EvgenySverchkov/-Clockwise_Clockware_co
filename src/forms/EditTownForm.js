import React, { useContext } from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";

import { updateTownInState } from "../store/townsManagement/actions";
import putDataToServer from "../components/AdminScreen/services/putDataToServer";

import EditForm from "./EditForm";

import Context from "../ContextComponent";

import { SERVERDOMAIN } from "../services/serverUrls";

const EditTownForm = ({ match, history }) => {
  const context = useContext(Context);
  const state = useSelector((state) => {
    return { townsArr: state.townReduser.towns };
  });
  const dispatch = useDispatch();

  function editTownHandler(e, newTownObj) {
    e.preventDefault();
    const targetElem = e.target;
    if (!targetElem.name.value.match(/^[A-Z]/)) {
      context.openWarningTooltip(
        "The string name must start with capital letter",
        targetElem.name.id
      );
      return false;
    }

    if (targetElem.name.value.match(/\d/)) {
      context.openWarningTooltip(
        "The string name must not contain numbers!",
        targetElem.name.id
      );
      return false;
    }
    if (!targetElem.name.value.match(/\b\w{3,}\b/)) {
      context.openWarningTooltip(
        "String name should not be shorter than 3 characters!",
        targetElem.name.id
      );
      return false;
    }

    if (targetElem.name.value.match(/\W/)) {
      context.openWarningTooltip(
        "String name should not containt punctuation characters and space character!",
        targetElem.name.id
      );
      return false;
    }

    putDataToServer(`${SERVERDOMAIN}/towns/put/${newTownObj.id}`, newTownObj)
      .then((data) => {
        if (data.success) {
          context.openSuccessWindowWithMsg(data.msg);
          dispatch(updateTownInState(data));
        } else {
          context.openErrorWindowWithMsg(data.msg);
        }
      })
      .catch((err) => context.openErrorWindowWithMsg(err));
  }
  return (
    <EditForm
      id={+match.params.id}
      handler={editTownHandler}
      arrFromState={state.townsArr}
    />
  );
};

EditTownForm.propTypes = {
  match: PropTypes.object,
  history: PropTypes.object,
};

export default EditTownForm;
