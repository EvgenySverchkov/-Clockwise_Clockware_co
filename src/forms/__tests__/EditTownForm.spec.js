import React from "react";

import {create, act} from "react-test-renderer";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";

import EditTownForm from "../EditTownForm";
import ContextComponent from "../../ContextComponent";

const mockStore = configureStore();

it("Test of <EditTownForm/> with mock props", async()=>{
    global.fetch = jest.fn(()=>Promise.resolve({json: ()=>Promise.resolve()}))
    const townId = 1;
    const store = mockStore({
        townReduser: {
            towns: [{id: townId, name: "Name"}]
        },
        clientTownsReduser: {
            townsInOrderFormIsLoad: false
        }
    });
    const mockProps = {
        match: {
            params: {
                id: townId
            }
        }
    }
    const mockContextValue = {
        closeWrningTooltip: jest.fn(),
        warningTooltipMsg: "Message",
        isOpenWarningTooltip: false,
        idForTooltip: 1
    }
    let component 
    await act(async()=>{
        component = await create(
            <Provider store = {store}>
                <ContextComponent.Provider value={mockContextValue}>
                    <EditTownForm {...mockProps}/>
                </ContextComponent.Provider>
            </Provider>)
    });
    expect(component.toJSON()).toMatchSnapshot();
});