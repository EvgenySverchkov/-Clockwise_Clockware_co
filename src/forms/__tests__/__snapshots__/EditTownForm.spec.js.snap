// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <EditTownForm/> with mock props 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter 
      name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
        value="Name"
      />
    </div>
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Edit"
    />
  </div>
</form>
`;
